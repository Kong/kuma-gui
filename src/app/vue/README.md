---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

