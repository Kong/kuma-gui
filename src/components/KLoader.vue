<template>
  <div class="loading-container">
    <img src="@/assets/images/kuma-loader-v1.gif">
    <div class="progress">
      <div
        :style="{ width: `${progress}%` }"
        class="progress-bar"
        role="progressbar"
      />
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      timer: null,
      progress: 10,
    }
  },

  mounted() {
    this.timer = setInterval(() => {
      if (this.progress >= 100) {
        clearInterval(this.timer)
        this.progress = 100
      }

      this.progress = Math.min(this.progress + Math.ceil(Math.random(10) * 30), 100)
    }, 150)
  },

  unmounted() {
    clearInterval(this.timer)
  },
}
</script>

<style lang="scss" scoped>
.loading-container {
  display: flex;
  max-width: 350px;
  align-items: center;
  flex-direction: column;

  .progress {
    margin-top: 1rem;
    width: 130%;
    background-color: #e6e6e6;

    .progress-bar {
      height: 5px;
      background-color: var(--LoaderProgressBarBG);
    }
  }
}
</style>
