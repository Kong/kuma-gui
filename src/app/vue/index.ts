import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
