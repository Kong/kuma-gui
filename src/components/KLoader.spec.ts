import { flushPromises } from '@vue/test-utils'
import { render, screen } from '@testing-library/vue'

import KLoader from './KLoader.vue'

jest.useFakeTimers()

describe('KLoader.vue', () => {
  it('renders snapshot at the beginning', () => {
    const { container } = render(KLoader)

    expect(container).toMatchSnapshot()
  })

  it('checks if width is max of 100% size', async () => {
    render(KLoader)

    jest.advanceTimersByTime(2500)

    await flushPromises()

    expect(screen.getByRole('progressbar')).toHaveAttribute('style', 'width: 100%;')
  })
})
