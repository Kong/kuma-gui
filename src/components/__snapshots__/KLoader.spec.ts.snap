// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KLoader.vue renders snapshot at the beginning 1`] = `
<div>
  <div
    class="loading-container"
  >
    <img
      src="../assets/images/kuma-loader-v1.gif?external"
    />
     
    <div
      class="progress"
    >
      <div
        class="progress-bar"
        role="progressbar"
        style="width: 10%;"
      />
    </div>
  </div>
</div>
`;
